### ISSUE TITLE
Request: support strict one-to-one association

### ISSUE THREAD

**Issue by anasyb** [balderdashy/waterline#1053]
I don't know how general is this bug, but here is the exact case I just run into:

- let's have an account model to store credentials
```
//Account model
//...
        // Relationships
        owner: {
            model: 'user'
        },
//...
    afterCreate: function(account, cb){
        User.create({account: account.id}).exec(console.log);
//...
```

- and obviously a user model
```
    //Relationships
    account: {
      model: 'account',
      via: 'owner'
    },
```

- now I create an account, only to be surprised to see in my mongo db:
```
/* USER collection */
{
    "_id" : ObjectId("55787acd6339c80100dde599"),
    "account" : ObjectId("55787acd6339c80100dde598"),//what??
}

/* ACCOUNT collection */
{
    "_id" : ObjectId("55787acd6339c80100dde598"),
    "username" : "...",
    "password" : "...",
    //no owner?!
}
```

- By product of that, if I query (via sails console) account I get no user associated with it.

**Comment by anasyb** [comment]
this seems to have nothing to do with `afterCreate()`...

**Comment by dmarcelino** [comment]
Hi @anasyb, what kind of association are you trying to create?

```js
/* user model */
   //Relationships
    account: {
      model: 'account',
      via: 'owner'
    },
```

`via` is used in *-to-many collections ([docs](https://github.com/balderdashy/waterline-docs/blob/master/associations.md#one-to-many-associations)), did you mean to use:

```js
    account: {
      collection: 'account',
      via: 'owner'
    },
```

**Comment by anasyb** [comment]
really, didn't know that we shouldn't use `via` with 1-to-1 :P
why is that?

**Comment by dmarcelino** [comment]
If you want one-to-one association it's probably not going to take any effect. Anyway, the reason why you are not getting an owner is:

> This will create a new `Pet` record with the `User` foreignKey set. It will allow you to query a `Pet`
and also retrieve their owners **but the `User` side of the association doesn't know about the `Pet`.
To ensure you can query both ways the `User` record will need to be updated with the new `Pet` record.**

https://github.com/balderdashy/waterline-docs/blob/master/associations.md#one-to-one-associations (scroll down to the second code example)

In the docs one-to-one association is really a one-way association and a proper one-to-one association has to be created/maintained manually.

**Comment by anasyb** [comment]
NOoOoO WAY!! 
so waterline provides magical way to sync 1-* and *-* associations, but that's not for 1-1.. I was taking this for granted (I have maybe 20+ associations, only one of them is one to one).
I assumed without thinking about it that sticking `via` there would make it two way association that gets stored in account side just like how one to many is.

**Comment by dmarcelino** [comment]
Yes way.

If you look underneath 1-1 is essentially a 1-* **without enforcing** that the other side is only associated with 1 record.

**Comment by anasyb** [comment]
Can we make this a feature request?

**Comment by devinivy** [comment]
Sure!  If you make a new issue with a short proposed spec (i.e., your idea using `via`), I'll tag it as a feature request needing community feedback.

**Comment by devinivy** [comment]
I'm repurposing this issue for now– changing the title from "associations break with Model.Create()".

**Comment by dmarcelino** [comment]
For reference there's a noteworthy discussion about this feature in #753.

**Comment by kevinob11** [comment]
Bringing my latest comment from #753 as you asked for a proposed spec:

In case where we have two models, person and pet, and person has a pet explicitly set on it when created. Then, similar to 1:n relationship config we can set an attribute on the pet that would imply a reference to the person that explicitly relates to that pet. The difference would be that we would use the model instead of collection key, to imply that this is 1:1 rather than 1:n. Example model config below. Then when either the person or the pet is loaded you can populate the related person or pet, from either side, even though the only explicit relation stored in the database is from the person to the pet.
```javascript
person: {
  attributes: {
    name: {
      type: 'text'
    },
    pet: {
      model: 'pet'
    }
  }
}

pet: {
  attributes: {
    name: {
      type: 'text'
    },
    person: {
      model: 'person',
      via: 'pet'
    }
  }
}
```

Look forward to feedback!

Thanks!

**Comment by sailsbot** [comment]
Thanks for posting, @anasyb.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by cjhall** [comment]
Just want to confirm whether this proposal has been abandoned before I go about implementing the two-way association through other means.  Is this parked?

**Comment by devinivy** [comment]
Good question– I defer to @tjwebb / @particlebanana.

**Comment by particlebanana** [comment]
@kevinob11's solution is pretty nice and is essentially the same thing as:

```javascript
person: {
  attributes: {
    name: {
      type: 'text'
    },
    pet: {
      model: 'pet'
    }
  }
}

pet: {
  attributes: {
    name: {
      type: 'text'
    },
    person: {
      collection: 'person',
      via: 'pet'
    }
  }
}

Pet.find().populate('person').exec(function(err, pets) {

  // Each pet will have { person: [{ id: 1, name: 'bob' }]
  // So we just need to set it to the first item in the array so its an object
  pets.forEach(function(pet) {
    pet.person = pet.person[0];
  });

});
```

That would mean we don't need to keep both sides updated. It wouldn't be trivial to change the join criteria in Waterline but we don't need to write much new code.

I wouldn't say it's been abandoned as much as put on hold for a while. We are trying to get a new release out that will fix bugs and possibly get deep populate out.

If you would like to see this moved forward you can submit a PR to the roadmap and add it. Thats how future feature requests will be accepted and scheduled.

**Comment by shaunc** [comment]
If `person.pet` has a unique index then this technique has the added benefit that there is no performance hit vs caching the key for the reverse relationship. A pet's own `id` would be just as efficient in the lookup.

vs @kevinob11 's spec -- there should be some way of telling which side ("source") has the key, right? Just "via" on "target" side isn't enough if there are multiple relationships from target to the same source.

**Comment by kevinob11** [comment]
I imagined that adding a `via` key to the one-to-one syntax would imply that the opposite side was the source side, but I agree that my model doesn't account for multiple relations from a single key or one-to-many relations. My spec was completely gauged to simple one-to-ones where I didn't want to maintain a reverse reference on create and update.

**Comment by shaunc** [comment]
A realistic use case involving 1-1 where "via" on one side wouldn't be enough would be ids are shared because of model inheritance. For instance we could imagine both a "Dog" model which was a Pet plus some other stuff, and a Parent which was a person plus some other stuff. Here both Pet & Dog would share ids as would Person/Parent. Waterline doesn't support directly, but perfectly possible (and IMHO reasonable but YMMV :)) to model this way in underlying data, especially if there are other applications accessing the data that do support that. In this case you'd want "via" on both sides to point at the base class. Upcast manually as again waterline doesn't support, but that shouldn't interfere with the relationships to the base classes.

The "virtual" keyword just means "the other side stores the ids for the relationship" and could be used for 1-1 in this case.

**Comment by kevinob11** [comment]
Fair enough, I'm all in on using a more explicit keyword such as "virtual".
